File: blockchain/blockchain/blockchain.cpp
#include "Blockchain.h"
#include <iostream>
#include <cmath>
#include <ctime>
#include <sstream>

Blockchain::Blockchain() {
    chain.push_back(createGenesisBlock());
}

Block Blockchain::createGenesisBlock() {
    Block genesis(0, "Genesis Block", "0", 2); // Start with difficulty 2 for parallel mining
    genesis.timestamp = std::time(nullptr);
    genesis.hash = genesis.calculateHash();
    genesis.miningTime = 0.0;  // Genesis block not mined
    return genesis;
}

bool Blockchain::addBlock(const Block& newBlock) {
    if (chain.empty()) {
        return false;
    }
    
    if (isValidNewBlock(newBlock, getLatestBlock())) {
        chain.push_back(newBlock);
        std::cout << "[BLOCKCHAIN] Block #" << newBlock.index << " added to chain. Hash: " 
                  << newBlock.hash.substr(0, 16) << "..." << std::endl;
        return true;
    }
    
    std::cout << "[BLOCKCHAIN] Invalid block rejected." << std::endl;
    return false;
}

bool Blockchain::isValidNewBlock(const Block& newBlock, const Block& previousBlock) const {
    // Check index
    if (previousBlock.index + 1 != newBlock.index) {
        std::cout << "[VALIDATION] Invalid index. Expected: " << (previousBlock.index + 1) 
                  << ", Got: " << newBlock.index << std::endl;
        return false;
    }
    
    // Check previous hash
    if (previousBlock.hash != newBlock.previousHash) {
        std::cout << "[VALIDATION] Invalid previous hash." << std::endl;
        return false;
    }
    
    // Check if the hash matches
    if (newBlock.calculateHash() != newBlock.hash) {
        std::cout << "[VALIDATION] Invalid hash calculation." << std::endl;
        return false;
    }
    
    // Check if hash meets difficulty requirement
    if (!hasValidHash(newBlock)) {
        std::cout << "[VALIDATION] Hash doesn't meet difficulty requirement." << std::endl;
        return false;
    }
    
    // Validate timestamp
    if (!isValidTimestamp(newBlock, previousBlock)) {
        std::cout << "[VALIDATION] Invalid timestamp." << std::endl;
        return false;
    }
    
    return true;
}

bool Blockchain::hasValidHash(const Block& block) const {
    std::string prefix(block.difficulty, '0');
    return block.hash.substr(0, block.difficulty) == prefix;
}

bool Blockchain::isValidTimestamp(const Block& newBlock, const Block& previousBlock) const {
    time_t currentTime = std::time(nullptr);
    
    // Block is valid if its timestamp is at most 1 minute greater than current time
    if (newBlock.timestamp > currentTime + 60) {
        return false;
    }
    
    // Block is valid if its timestamp is at most 1 minute less than previous block's timestamp
    if (newBlock.timestamp < previousBlock.timestamp - 60) {
        return false;
    }
    
    return true;
}

bool Blockchain::isValidChain(const std::vector<Block>& chainToValidate) const {
    if (chainToValidate.empty()) {
        return false;
    }
    
    // Check genesis block
    const Block& genesisBlock = chainToValidate[0];
    if (genesisBlock.index != 0 || genesisBlock.previousHash != "0") {
        return false;
    }
    
    // Validate each block
    for (size_t i = 1; i < chainToValidate.size(); i++) {
        if (!isValidNewBlock(chainToValidate[i], chainToValidate[i - 1])) {
            return false;
        }
    }
    
    return true;
}

bool Blockchain::replaceChain(const std::vector<Block>& newChain) {
    if (!isValidChain(newChain)) {
        std::cout << "[BLOCKCHAIN] Received chain is invalid." << std::endl;
        return false;
    }
    
    double currentDifficulty = getCumulativeDifficulty(chain);
    double newDifficulty = getCumulativeDifficulty(newChain);
    
    if (newDifficulty > currentDifficulty) {
        std::cout << "[BLOCKCHAIN] Replacing chain. New cumulative difficulty: " 
                  << newDifficulty << " > " << currentDifficulty << std::endl;
        chain = newChain;
        return true;
    }
    
    std::cout << "[BLOCKCHAIN] Received chain has lower cumulative difficulty." << std::endl;
    return false;
}

double Blockchain::getCumulativeDifficulty(const std::vector<Block>& chainToCheck) const {
    double total = 0;
    for (const auto& block : chainToCheck) {
        total += std::pow(2, block.difficulty);
    }
    return total;
}

int Blockchain::getDifficulty() const {
    return calculateDifficulty(false); // No logging for backward compatibility
}

int Blockchain::getDifficulty(int mpiRank) const {
    return calculateDifficulty(mpiRank == 0); // Only log on rank 0
}

int Blockchain::calculateDifficulty(bool enableLogging) const {
    const Block& latestBlock = getLatestBlock();
    
    // If we haven't reached the adjustment interval, keep the same difficulty
    if (latestBlock.index % DIFFICULTY_ADJUSTMENT_INTERVAL != 0 || latestBlock.index == 0) {
        return latestBlock.difficulty;
    }
    
    // Get the adjustment block
    const Block& adjustmentBlock = chain[chain.size() - DIFFICULTY_ADJUSTMENT_INTERVAL];
    
    // Calculate expected and actual MINING time (not wall-clock time)
    double expectedTime = static_cast<double>(BLOCK_GENERATION_INTERVAL) * static_cast<double>(DIFFICULTY_ADJUSTMENT_INTERVAL);
    
    // Sum actual mining time for the last DIFFICULTY_ADJUSTMENT_INTERVAL blocks
    double actualMiningTime = 0.0;
    for (size_t i = chain.size() - DIFFICULTY_ADJUSTMENT_INTERVAL; i < chain.size(); i++) {
        actualMiningTime += chain[i].miningTime;
    }
    
    // Calculate adjustment factor
    double timeFactor = actualMiningTime / expectedTime;
    
    int newDiff = adjustmentBlock.difficulty;
    
    // Adjust difficulty based on mining time (only ±1 at a time for smoother adjustment)
    if (timeFactor < 0.5) {
        // Mining was 2x faster than target - increase difficulty
        newDiff = adjustmentBlock.difficulty + 1;
        if (enableLogging) {
            std::cout << "[DIFFICULTY] Mining too fast (" << actualMiningTime << "s vs " << expectedTime 
                      << "s target). Increasing: " << adjustmentBlock.difficulty << " -> " << newDiff << std::endl;
        }
    } 
    else if (timeFactor > 2.0) {
        // Mining was 2x slower than target - decrease difficulty
        newDiff = adjustmentBlock.difficulty - 1;
        if (newDiff < 1) newDiff = 1; // Minimum difficulty of 1
        if (enableLogging) {
            std::cout << "[DIFFICULTY] Mining too slow (" << actualMiningTime << "s vs " << expectedTime 
                      << "s target). Decreasing: " << adjustmentBlock.difficulty << " -> " << newDiff << std::endl;
        }
    }
    else {
        if (enableLogging) {
            std::cout << "[DIFFICULTY] Mining time acceptable (" << actualMiningTime << "s vs " << expectedTime 
                      << "s target). Keeping difficulty at " << adjustmentBlock.difficulty << std::endl;
        }
    }
    
    return newDiff;
}

std::string Blockchain::toJson() const {
    std::stringstream ss;
    ss << "[";
    for (size_t i = 0; i < chain.size(); i++) {
        ss << chain[i].toJson();
        if (i < chain.size() - 1) {
            ss << ",";
        }
    }
    ss << "]";
    return ss.str();
}

std::vector<Block> Blockchain::chainFromJson(const std::string& json) {
    std::vector<Block> result;
    
    // Simple JSON array parsing
    size_t pos = 1; // Skip opening '['
    while (pos < json.length()) {
        size_t start = json.find('{', pos);
        if (start == std::string::npos) break;
        
        size_t end = json.find('}', start);
        if (end == std::string::npos) break;
        
        std::string blockJson = json.substr(start, end - start + 1);
        result.push_back(Block::fromJson(blockJson));
        
        pos = end + 1;
    }
    
    return result;
}

File: web-app/backend/utils/blockchain.js
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

const BLOCKCHAIN_FILE = path.join(__dirname, '../data/blockchain.json');
const DIFFICULTY = 3; // Number of leading zeros required (3 = ~1-2 seconds, 4 = ~10-20 seconds)
const BLOCK_GENERATION_INTERVAL = 10; // Target: 10 seconds per block
const DIFFICULTY_ADJUSTMENT_INTERVAL = 2; // Adjust difficulty every 2 blocks

/**
 * Simple Blockchain Implementation for Educational Purposes
 * Demonstrates: Hashing, Proof-of-Work, Chain Validation, Difficulty Adjustment
 */

class Block {
    constructor(index, data, previousHash, difficulty = DIFFICULTY) {
        this.index = index;
        this.data = data;
        this.timestamp = Date.now();
        this.previousHash = previousHash;
        this.difficulty = difficulty;
        this.nonce = 0;
        this.hash = '';
    }

    /**
     * Calculate SHA-256 hash of the block
     */
    calculateHash() {
        const dataString = `${this.index}${this.data}${this.timestamp}${this.previousHash}${this.difficulty}${this.nonce}`;
        return crypto.createHash('sha256').update(dataString).digest('hex');
    }

    /**
     * Check if hash meets the difficulty requirement (starts with N zeros)
     */
    hashMeetsDifficulty(hash) {
        const prefix = '0'.repeat(this.difficulty);
        return hash.startsWith(prefix);
    }

    /**
     * Mine the block - find a nonce that produces a hash meeting the difficulty
     * This is the "proof-of-work" - computationally expensive to find, easy to verify
     */
    mine() {
        console.log(`[Blockchain] Mining block #${this.index} with difficulty ${this.difficulty}...`);
        const startTime = Date.now();
        
        while (!this.hashMeetsDifficulty(this.hash)) {
            this.nonce++;
            this.hash = this.calculateHash();
            
            // Log progress every 10000 attempts
            if (this.nonce % 10000 === 0) {
                process.stdout.write(`\r[Blockchain] Attempts: ${this.nonce.toLocaleString()}...`);
            }
        }
        
        const miningTime = ((Date.now() - startTime) / 1000).toFixed(2);
        console.log(`\n[Blockchain] ✓ Block mined in ${miningTime}s (nonce: ${this.nonce.toLocaleString()})`);
        return this;
    }

    /**
     * Validate block structure and hash
     */
    isValid(previousBlock) {
        // Check hash calculation
        if (this.calculateHash() !== this.hash) {
            console.error('[Blockchain] Invalid: Hash calculation mismatch');
            return false;
        }

        // Check difficulty requirement
        if (!this.hashMeetsDifficulty(this.hash)) {
            console.error('[Blockchain] Invalid: Hash does not meet difficulty requirement');
            return false;
        }

        // Check previous hash link (except for genesis block)
        if (previousBlock && this.previousHash !== previousBlock.hash) {
            console.error('[Blockchain] Invalid: Previous hash mismatch');
            return false;
        }

        // Check index continuity
        if (previousBlock && this.index !== previousBlock.index + 1) {
            console.error('[Blockchain] Invalid: Index not continuous');
            return false;
        }

        return true;
    }
}

class Blockchain {
    constructor() {
        this.chain = [];
        this.loadChain();
        
        // Create genesis block if chain is empty
        if (this.chain.length === 0) {
            this.createGenesisBlock();
        }
    }

    /**
     * Create the first block in the chain (Genesis Block)
     */
    createGenesisBlock() {
        const genesis = new Block(0, 'Genesis Block - Apopulis News Blockchain', '0', DIFFICULTY);
        genesis.hash = genesis.calculateHash();
        // Genesis block doesn't need mining (or use difficulty 0)
        genesis.difficulty = 0;
        this.chain.push(genesis);
        this.saveChain();
        console.log('[Blockchain] Genesis block created');
    }

    /**
     * Get the latest block in the chain
     */
    getLatestBlock() {
        return this.chain[this.chain.length - 1];
    }

    /**
     * Calculate current difficulty based on block generation time
     * Adjusts difficulty to maintain ~10 seconds per block
     */
    getDifficulty() {
        if (this.chain.length < DIFFICULTY_ADJUSTMENT_INTERVAL + 1) {
            return DIFFICULTY;
        }

        const latestBlock = this.getLatestBlock();
        const previousAdjustmentBlock = this.chain[this.chain.length - DIFFICULTY_ADJUSTMENT_INTERVAL - 1];
        
        const timeExpected = BLOCK_GENERATION_INTERVAL * DIFFICULTY_ADJUSTMENT_INTERVAL;
        const timeTaken = (latestBlock.timestamp - previousAdjustmentBlock.timestamp) / 1000;
        
        const timeFactor = timeTaken / timeExpected;
        
        let newDifficulty = previousAdjustmentBlock.difficulty;
        
        if (timeFactor < 0.4) {
            newDifficulty += 2; // Very fast - increase difficulty significantly
        } else if (timeFactor < 0.75) {
            newDifficulty += 1; // Fast - increase difficulty
        } else if (timeFactor > 2.5) {
            newDifficulty -= 2; // Very slow - decrease difficulty significantly
        } else if (timeFactor > 1.33) {
            newDifficulty -= 1; // Slow - decrease difficulty
        }
        
        // Keep difficulty between 1 and 6 for reasonable mining times
        return Math.max(1, Math.min(6, newDifficulty));
    }

    /**
     * Add a new block to the blockchain
     * @param {string} data - The data to store in the block (e.g., news item ID)
     * @returns {Block} The newly mined block
     */
    addBlock(data) {
        const previousBlock = this.getLatestBlock();
        const difficulty = this.getDifficulty();
        
        const newBlock = new Block(
            previousBlock.index + 1,
            data,
            previousBlock.hash,
            difficulty
        );

        // Mine the block (proof-of-work)
        newBlock.mine();

        // Validate before adding
        if (newBlock.isValid(previousBlock)) {
            this.chain.push(newBlock);
            this.saveChain();
            console.log(`[Blockchain] Block #${newBlock.index} added to chain`);
            return newBlock;
        } else {
            throw new Error('Invalid block cannot be added to chain');
        }
    }

    /**
     * Validate the entire blockchain
     */
    isValidChain() {
        // Check genesis block
        const genesis = this.chain[0];
        if (genesis.index !== 0 || genesis.previousHash !== '0') {
            console.error('[Blockchain] Invalid genesis block');
            return false;
        }

        // Check each block
        for (let i = 1; i < this.chain.length; i++) {
            const currentBlock = this.chain[i];
            const previousBlock = this.chain[i - 1];

            if (!currentBlock.isValid(previousBlock)) {
                console.error(`[Blockchain] Invalid block at index ${i}`);
                return false;
            }
        }

        return true;
    }

    /**
     * Get blockchain statistics
     */
    getStats() {
        return {
            length: this.chain.length,
            latestBlock: {
                index: this.getLatestBlock().index,
                hash: this.getLatestBlock().hash.substring(0, 16) + '...',
                difficulty: this.getLatestBlock().difficulty,
                timestamp: new Date(this.getLatestBlock().timestamp).toISOString()
            },
            currentDifficulty: this.getDifficulty(),
            isValid: this.isValidChain()
        };
    }

    /**
     * Load blockchain from JSON file
     */
    loadChain() {
        try {
            if (fs.existsSync(BLOCKCHAIN_FILE)) {
                const data = fs.readFileSync(BLOCKCHAIN_FILE, 'utf8');
                this.chain = JSON.parse(data);
                console.log(`[Blockchain] Loaded ${this.chain.length} blocks from file`);
            }
        } catch (error) {
            console.error('[Blockchain] Error loading chain:', error.message);
            this.chain = [];
        }
    }

    /**
     * Save blockchain to JSON file
     */
    saveChain() {
        try {
            // Ensure data directory exists
            const dataDir = path.dirname(BLOCKCHAIN_FILE);
            if (!fs.existsSync(dataDir)) {
                fs.mkdirSync(dataDir, { recursive: true });
            }

            fs.writeFileSync(BLOCKCHAIN_FILE, JSON.stringify(this.chain, null, 2));
        } catch (error) {
            console.error('[Blockchain] Error saving chain:', error.message);
        }
    }
}

// Export singleton instance
const blockchain = new Blockchain();

module.exports = blockchain;
